### PR TITLE
Ticket 3

### PR DESCRIPTION
**Pull request by TheRealImrano**
completed ticket-3 and attempting a pull request. Since I was already working on this before completing ticket-2, I am going to have a merge conflict, so I may need help with that. Many thanks.

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        .then((response)=>{
+            const filePath = './endpoints.json';
+            readJsonFile(filePath)
+            .then(result => {
+                expect(response.body).toEqual(result);
+            })
+            // expect(response.body).toEqual(readJsonFile(filePath));
+        })
+    })
+})
```
Comment (leekli) [inline __tests__/app.test.js]: Couple of bits:

- Good approach to the testing, which is do the request to `/api` and then check the response matches what's in the `endpoints.json` file.
- In the test description, it's worth adding the expected response code
- I can see what you're doing with the readJsonFile, I'd argue it's slightly over engineered, and for this you could just use `fs.readFile` to get the file contents then just use that
- There's a rogue console log on line 46 which needs removing

File: app.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    const filePath = './endpoints.json';
+
+    readJsonFile(filePath)
+    .then(data => {
+        res.status(200).send(data);
+    })
+    .catch(err => {
+        next(err)
+    });
+});
```
Comment (leekli) [inline app.js]: On this:

- Great work setting up the `/api` endpoint
- The login within it should sit in a controller, not directly in here, so please set up this into a controller.
- Same feedback here regarding your `readJsonFile` function, it's not really needed, we have a battle tested function already available to us from `fs` and `fs.readFile`.
- Great to see a catch block to deal with an error

File: app.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        console.log(err.status, ':', err.message);
+      res.status(err.status).send({ msg: err.msg });
+    } else next(err);
+  });
+
+app.use((err, req, res, next) => {
+    console.log('server error');
+    console.error(err.stack);
+    res.status(500).json({ error: "Internal Server Error" });
+});
```
Comment (leekli) [inline app.js]: Lovely to see you've started to error handling, well done!

The error response is being dealt with in the `res` to it, so I'd encourage you to take the console logs out now

File: utils/jsonReader.js [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+exports.readJsonFile = (filePath) => {
+    return fs.readFile(filePath, 'utf-8')
+        .then(jsonData => JSON.parse(jsonData))
+        .catch(err => {
+            return Promise.reject({
+                status: 500,
+                message: 'Error reading JSON file',
+            });
+        });
+}
```
Comment (leekli) [inline utils/jsonReader.js:12]: See my previous feedback on this

**Review by leekli — CHANGES_REQUESTED** [review]
Hey, well done so far :)

I've left some comments and feedback throughout for you to review.

Main changes:

- I like where you're going with your own `readJsonFile` util function but it isn't needed, `fs.readFile` is a reusable function already which we should make use of.
- Your logic for the `/api` endpoint in `app.js` needs extracting out into a controller

Please have a look at making these changes and then resubmit the pull review request once finished, for another review from us.

**Comment by TheRealImrano** [comment]
thanks for the feedback. Ill implement right away

**Review by TheRealImrano — COMMENTED** [review]
Made amends to ticket 3; merge conflict needs resolving

**Review by leekli — APPROVED** [review]
Hi Ali, thanks for making the requested changes.

These look great, you are free to merge this branch.

regarding merge conflicts - it depends what the conflicts are, I recommend you have a research on how to deal with them, this youTube video looks informative: https://www.youtube.com/watch?v=QmKdodJU-js

If still stuck with a merge conflict, feel free to nchelp us and we'll get to you when we're able to.

**Comment by TheRealImrano** [comment]
I'll take a look at the video. Many thanks Lee!

**Pull request merged**